Add a document describing key revocation

mnm678 opened the pull request:
This document lays out some of the options for key revocation that have been discussed. These options might eventually fit better as part of the key management document, but are posted separately for the sake of discussion.

mnm678 committed:
Add a document describing key revocation.

This document may eventually be part of the key management
requirements. It describes a few common mechanisms for
key revocation.

Signed-off-by: Marina Moore <mnm678@gmail.com>

joshuagl commented on key-revocation.md:
```diff
@@ -0,0 +1,33 @@
+
+One of the goals of Notary v2 is to build in solutions for key revocation that are easy to use and ensure that users will always use non-compromised keys. This document discusses some potential mechanisms for key revocation.
+
+In existing systems, there are three main approaches to key revocation: automatic revocation through key expiration, key revocation lists, and distribution of trusted keys. I discuss some of the benefits and pitfalls of each of these techniques, and how some of these techniques are combined to provide a wholistic approach to key revocation in TUF.
```
```suggestion
In existing systems, there are three main approaches to key revocation: automatic revocation through key expiration, key revocation lists, and distribution of trusted keys. I discuss some of the benefits and pitfalls of each of these techniques, and how some of these techniques are combined to provide a holistic approach to key revocation in TUF.
```

joshuagl commented on key-revocation.md:
```diff
@@ -0,0 +1,33 @@
+
+## Distribution of trusted keys
+
+Instead of distributing untrusted keys, this method distributes a list of currently trusted keys. If a key needs to be revoked, it is removed from the list of trusted keys. This technique as the added benefit of ensuring that users have access to the new trusted key as soon as they learn of a revocation.
```
```suggestion
Instead of distributing untrusted keys, this method distributes a list of currently trusted keys. If a key needs to be revoked, it is removed from the list of trusted keys. This technique has the added benefit of ensuring that users have access to the new trusted key as soon as they learn of a revocation.
```

sudo-bmitch commented on key-revocation.md:
```diff
@@ -0,0 +1,33 @@
+
+However, the user must be able to ensure that the key revocation list is accurate and up to date. If an attacker is able to replay an old revocation list, the user may continue to trust compromised keys. Therefore the distribution of the key revocation list must allow the user to verify authenticity and timeliness.
+
+Also, for security reasons, keys cannot be removed from a key revocation list, so the list will grow larger and larger over time and may eventually have a noticeable bandwidth impact.
```
To mitigate the risk of ever growing revocation list, there can be a combination, a longish expiration time on keys that can be revoked. Then a revoked key can be removed from the list after it eventually expires.

sudo-bmitch reviewed (commented):
Each of these options has pros and cons. Thinking through them:

Key Expiration: this has the advantage of being automatically enforced, even for disconnected environments. I'd question if this means everything the key has previously signed would need to be resigned. Fairly certain the answer is yes (otherwise an attacker could sign a malicious image for 10 years when they breach a key that only has 2 hours left until the certificate on the key expires). That would result in lots of re-signing of old images for every key rotation. Perhaps that could be made easier by having a single signature for a list of images (digests), rather than a separate signature for each individual image.

Revocation lists: while it's convenient that this has an immediate affect when the revocation is published, I'm seeing multiple downsides. In disconnected environments, that query may fail, or it may be sent to a mirror server the client in that disconnected environment is told to trust. A stale mirror in the disconnected environment could be used to send malicious images, though in those cases it's the client intentionally indicating they want to trust a mirror that shouldn't have been trusted. The more concerning scenario to me are the devices with access to the public internet using the upstream revocation list. What do we do when access to that revocation list goes down? Do we fail insecure and potentially allowing a vulnerability, or fail secure and cause an outage. Last year's Apple scenario showed we can have the worst of both, where the revocation server could be extremely slow to eventually timeout on the response.

For the TUF scenario, I think we want to explore what it would look like for the root key to be eventually expired (with a relatively long lifetime). And with short lifetimes on the timestamp signing, what does that look like for mirrors and popular registries that want to push as much out to CDN's as possible.

And it's bigger than this document scoped, but we also need to explore what key distribution looks like with v2. If we are avoiding TOFU by having clients explicitly trusting a root key for the organization, how is that root key first deployed, how does it get rotated, and if we do in-band rotation, can we trust that chain of rotated root keys that lead back to one or more now expired keys.

mnm678 committed:
Clarify and fix typos

Signed-off-by: Marina Moore <mnm678@gmail.com>

mnm678 committed:
Add pros and cons

Add an initial list of pros and cons for each technique
and add a few clarifications

Signed-off-by: Marina Moore <mnm678@gmail.com>

mnm678 committed:
Generalize description of key revocation in TUF

Signed-off-by: Marina Moore <mnm678@gmail.com>

mnm678 committed:
Update key revocation option

Update the combined key revocation option to remove references
to TUF and more generically describe the way it allows for both
explicit and implicit key revocation.

Thanks to @sudo-bmitch for wording suggestions.

Signed-off-by: Marina Moore <mnm678@gmail.com>

gokarnm commented on key-revocation.md:
```diff
@@ -0,0 +1,75 @@
+
+## Key Expiration
+
+Adding an expiration time to every key allows keys to automatically be revoked after a certain period of time. The expiration time is usually included with the key so that it is easy for users to find. This technique does not require any action from the key holder, and ensures that users will have to refresh their trusted keys before those keys expire.
```
It might help to clarify that the key expiry (metadata) needs to be signed by a issuing key that the client trusts.

gokarnm commented on key-revocation.md:
```diff
@@ -0,0 +1,75 @@
+
+Cons:
+* Keys can't be revoked before expiration
+* Artifacts must be re-signed after expiration
```
Timestamping is an option that allows use of signed artifacts after the key expires.

gokarnm commented on key-revocation.md:
```diff
@@ -0,0 +1,75 @@
+* Artifacts must be re-signed after expiration
+
+
+## Key revocation lists (Deny lists)
```
This section and next is similar to [Allowlist](https://hackmd.io/2rc07WwHTdmPjpgoMbvYfA?view#Allowlist-in-Repository) and [Denylist in Repository](https://hackmd.io/2rc07WwHTdmPjpgoMbvYfA?view#Denylist-in-Repository), the challenge is synchronizing deny lists in multi-registry scenarios. Also we don't cover artifact level revocation in this doc, and specify where allow/deny lists will be stored.

mnm678 replied:
I added a note about synchronization. I purposefully left out artifact level revocation in this document for now, but we can combine those discussions is a later draft.

gokarnm commented on threatmodel.md:
```diff
@@ -10,3 +10,10 @@ It is assumed that an attacker may perform one or more the following actions:
 
 While it is not always possible to protect against all scenarios, the system should to the extent possible mitigate and/or reduce the damage caused by a successful attack, detect the occurrence of an attack and notify appropriate parties, yet remain usable for parties operating the system.  Furthermore, the system should recover from successful attacks in a way that presents low operational overh ...
 
+Attacker Goals:
```
Should these be reworded to be less generic terminology, and targeted for artifact registry and consumers? Also is this intended to be an initial version? I think the final threat model and analysis would be detailed.

mnm678 replied:
It looks like these are duplicated in #35, and they are a bit out of scope for this pr, so I'll remove them here and try to address your comment there

mnm678 committed:
Add clarifications

Signed-off-by: Marina Moore <mnm678@gmail.com>

mnm678 committed:
remove attacker goals from this pr

Signed-off-by: Marina Moore <mnm678@gmail.com>

mnm678 committed:
remove references to solution in the intro for this overview document

Signed-off-by: Marina Moore <mnm678@gmail.com>

SteveLasker commented on key-revocation.md:
```diff
@@ -0,0 +1,76 @@
+
+However, the user must be able to ensure that the key revocation list is accurate and up to date. If an attacker is able to replay an old revocation list, or show different versions to different registries, the user may continue to trust compromised keys. Therefore the distribution of the key revocation list must allow the user to verify authenticity and timeliness.
+
+Also, for security reasons, keys cannot be removed from a key revocation list, so the list will grow larger and larger over time and may eventually have a noticeable bandwidth impact, although this can be mitigated by combining key revocation lists with keys that expire.
```
Is this true, that once in a list, it's never removed? Or, can keys that are known to have expired be removed at a later date? Perhaps 50% longer than the life of the key or something. It does seem like a non-scalable solution that needs mitigation.

mnm678 replied:
This could be combined with strongly enforced key expiration to allow the keys to eventually be deleted.

SteveLasker commented on key-revocation.md:
```diff
@@ -0,0 +1,76 @@
+
+This method allows signers to have relatively long lived keys, to simplify their workflow and avoid needing to resign the artifacts themselves, while enabling timely revoking of the signing key or a single artifact signature.
+
+For efficiency, a meta-artifact can be created and maintained, containing references to a collection currently signed artifacts. And the short lived signature can be created for this single artifact, rather than every artifact individually.
```
How would this interact with individual artifacts moving within and across registries?

mnm678 replied:
it depends on the implementation, but the meta-artifact would have to be updated as the collection changes.

SteveLasker commented on key-revocation.md:
```diff
@@ -0,0 +1,76 @@
+Cons:
+* Requires maintenance of an automated system to refresh short lived signatures
+* A root key compromise requires updating all signers, clients, and signatures on the artifacts
+* Updating short lived signatures on a large number of artifacts may encounter scaling challenges and loses some of the caching efficiencies of content addressable storage in registries
```
Can we add a con here for the meta-artifact collection of keys needs to somehow be parseable for individual artifact movement within and across registries?

mnm678 replied:
The meta-artifact is an optional efficiency feature, so I added the con to the paragraph describing the feature.

SteveLasker reviewed (commented):
If we can add a con to the meta-artifact section, addressing the multi-registry challenge for moving individual artifacts

mnm678 committed:
Add description of meta-artifact downsides for artifact movement

Signed-off-by: Marina Moore <mnm678@gmail.com>

SteveLasker reviewed (approved):
LGTM
